cli-plugins/manager/plugin.go

```diff
@@ -0,0 +1,145 @@
+package manager
+
+import (
+	"encoding/json"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+var (
+	pluginNameRe = regexp.MustCompile("^[a-z][a-z0-9]*$")
+)
+
+// PluginError is set as Plugin.Err by NewPlugin if the plugin
+// candidate fails one of the candidate tests. This exists primarily
+// to implement encoding.TextMarshaller such that rendering a plugin as JSON
+// (e.g. for `docker info -f '{{json .CLIPlugins}}'`) renders the Err
+// field as a useful string and not just `{}`. See
+// https://github.com/golang/go/issues/10748 for some discussion
+// around why the builtin error type doesn't implement this.
+type PluginError struct {
+	cause error
+}
+
+// Error satisfies the core error interface for PluginError.
+func (e *PluginError) Error() string {
+	return e.cause.Error()
+}
+
+// Cause satisfies the errors.causer interface for PluginError.
+func (e *PluginError) Cause() error {
+	return e.cause
+}
+
+// MarshalText marshalls the PluginError into a textual form.
+func (e *PluginError) MarshalText() (text []byte, err error) {
+	return []byte(e.cause.Error()), nil
+}
+
+// WrapAsPluginError wraps an error in a PluginError with an
+// additional message, analogous to errors.Wrapf.
+func WrapAsPluginError(err error, msg string) error {
+	return &PluginError{cause: errors.Wrap(err, msg)}
+}
+
+// NewPluginError creates a new PluginError, analogous to
+// errors.Errorf.
+func NewPluginError(msg string, args ...interface{}) error {
+	return &PluginError{cause: errors.Errorf(msg, args...)}
+}
+
+// Plugin represents a potential plugin with all it's metadata.
+type Plugin struct {
```

Comment:
nit: Move this type up in the file, as it is certainly the main type here users will read as a reference?